src/Uno.UWP/Globalization/ApplicationLanguages.cs

```diff
@@ -18,11 +19,42 @@ static ApplicationLanguages()
 
 		public static string PrimaryLanguageOverride
 		{
-			get => _primaryLanguageOverride;
+			get
+			{
+				if (_primaryLanguageOverride.Length == 0 &&
+					ApplicationData.Current.LocalSettings.Values.TryGetValue(PrimaryLanguageOverrideSettingKey, out var savedValue) &&
+					savedValue is string stringSavedValue)
+				{
+					_primaryLanguageOverride = stringSavedValue;
+				}
+
+				return _primaryLanguageOverride;
+			}
 			set
 			{
+				if (value is null)
+				{
+					throw new ArgumentNullException(nameof(value), "Value cannot be null.");
+				}
+
+				try
+				{
+					// This doesn't match UWP for edge cases. In UWP:
+					// Windows.Globalization.ApplicationLanguages.PrimaryLanguageOverride = "fr-abcdefgh"  // This succeeds.
+					// Windows.Globalization.ApplicationLanguages.PrimaryLanguageOverride = "fr-abcdefghe" // This fails.
+					// However, new CultureInfo(..) fails in both cases.
+					// These are edge cases anyway.
+					_ = new CultureInfo(value);
+				}
+				catch (CultureNotFoundException)
+				{
+					throw new ArgumentException("Value does not fall within the expected range.");
+				}
```

Comment:
@jeromelaban In UWP app `new CultureInfo("fr-Latn-CA")` doesn't throw, but when setting LanguageOverride to this value, it magically turns into `"fr-CA"` (I don't know what's logic behind this).

However things are even worse since Mono runtime throws an exception when "fr-Latn-CA" is passed to CultureInfo ctor. I'm completely unsure how should we move forward here.

Reply:
We may not want for the language property override to fail, that would make an app unusable if an user's current culture matches this throwing construct.

We may need to re-create a simpler culture through string splitting (e.g. if there are two dashes, we take first and last parts, then first only so it does not fail). I know @MartinZikmund had a problem with his own culture which has three parts.

Reply:
@jeromelaban Taking another look, I think the logic in UWP is more complex than this. In UWP, `ApplicationLanguages.PrimaryLanguageOverride = "az-Latn-AZ";` sets it to `az-Latn` (third part ignored), however, for `fr-Latn-CA` the value is set to `fr-CA` (second part ignored).

My best guess is for the three-part cultures case:

- Check if `FirstPart-ThirdPart` exists in `CultureInfo.GetCultures(CultureTypes.SpecificCultures)`, use that. (This is like the `fr-Latn-CA` case).
- If not, use `FirstPart-SecondPart` (This is like the `az-Latn-AZ` part).
- In all cases, throw an exception if any part is greater than 4 characters

For two-part cultures, rely on `new CultureInfo(...)` as it's now.

@jeromelaban Does that sounds good and/or correct?

Reply:
@Youssef1313 lets try this :)

Reply:
There is even more magic than that. `zh-Hans-CN` is set as-is. So we have three different 3-part cultures, which 3 different behaviors on UWP.

I'm not really sure what to do here. Ignoring the validation completely will cause problems when we try to set this value as `CurrentCulture` and `CurrentUICulture`. Any suggestions?

Reply:
Let's handle the scenarios we know of, and we'll adjust as we find out more.

Reply:
@jeromelaban Actually I think there is no need for custom logic here. It seems like `PrimaryLanguageOverride` follows whatever `Language` class do in UWP.

The `Language` class implementation looks far away from being implemented correctly in Uno, and relying on it is certainly the correct way to go here. But it really needs to get fixed first. Currently relying on `Language` API will really fail in many scenarios that we even have tests for.